### ISSUE TITLE
Security and UserAllowedIP

### ISSUE THREAD

**Issue by uldiseihenbergs** [Ylianst/MeshCentral#1119]
I'm seeing something similar, like #1069 :
i'm using UserAllowedIP to allow from LAN side only.
i'm trying to add Linux agent outside my network, browsing to /meshagents says "Unauthorized", but /meshagents?script=1, it downloads meshinstall.sh and runs it, but fails in some next download with "Unauthorized" again. IMHO, /meshagents?script=1 should not download too in such configuration..

may i suggest:
- when you open Add mesh agent window, avoid checking for UserAllowedIP for some time to allow add agents outside UserAllowedIPs?
- add support for DNS names under UserAllowedIP. it would be useful when dynamic DNS is used and i want to allow user from outside my LAN, but agents still connect from anywhere.

Thanks for your hard work!
Uldis

**Comment by Ylianst** [comment]
So, this is a bit tricky. I don't want to "avoid checking UserAllowedIP" for some time because it's not secure and people can download the agent install command and run it anytime in the future. The person getting the agent installation command may put it in a USB and run it much later. I will accept other suggestions, but not this one.

For DNS names until UserAllowedIP, MeshCentral could do a reverse lookup, but in general users don't generally browser from computers that have DNS names. Would this really be useful? Also, how often would the DNS reverse lookup occur?

Apologies for kicking back the suggestions, but I am open to changes. Note that there is also a "AgentAllowedIP" that works the same at "UserAllowedIP" but for agents.

**Comment by uldiseihenbergs** [comment]
huh, maybe you should check for AgentAllowedIP on agent install instead on UserAllowedIP? 
i have no AgentAllowedIP at this time in my config, but i have UserAllowedIP. Today i added one windows agent successfully, but can't add Linux agent because Linux agent install tries to download script. i think, it shoul work equally Linux vs Windows..
and according to DDNS - in these Covid days lot of us are working from home, and i'm on mobile 4G LTE connection with DDNS, while my Meshcentral is in my office. VPN should always work, but i prefer DDNS, so yes it would be fine if MC could do reverse DNS lookup, how often - 10 minutes will be sufficient.
Thanks,
Uldis

**Comment by Ylianst** [comment]
Every 10 minutes! I don't think DDNS is going to be a priority, it's a lot of work and likely not going to be used much. If someone builds it and does a pull request for it, I will accept it. Maybe look for specialized software you can put in front of MeshCentral that does this?

Right now, AgentAllowedIP should only affect agent connections, not installation. I will take a look again at Linux installation and make sure it's not blocked by UserAllowedIPs.

**Comment by uldiseihenbergs** [comment]
DDNS: i agree, it's not a very high priority job. i'm on VPN now. if it takes much of work, then forget about it. :)

**Comment by Ylianst** [comment]
I just tested Linux agent installation using "LoginKey" and it works for me. Note that the login key is added to the URL in the script, notice the &key=abc below.

```
wget "https://xxxx.meshcentral.com/meshagents?script=1&key=abc" --no-proxy --no-check-certificate -O ./meshinstall.sh && chmod 755 ./meshinstall.sh && sudo ./meshinstall.sh https://xxx.meshcentral.com 'xxxxxxxx' || ./meshinstall.sh https://xxx.meshcentral.com 'xxxxxxxx'
```

If there is any issues with agent installation when the LoginKey is used, please post an exact copy the error. Ideally the exact URL that failed to load. Thanks.

**Comment by uldiseihenbergs** [comment]
LoginKey? i have no such available in my case (why it downloads script twice? see below).

i just click on "Invite" link, a page opens, and on Linux page i get a following link:
`(wget "https://mesh.mydomain.lv/meshagents?script=1" --no-check-certificate -O ./meshinstall.sh || wget "https://mesh.mydomain.lv/meshagents?script=1" --no-proxy --no-check-certificate -O ./meshinstall.sh) && chmod 755 ./meshinstall.sh && sudo ./meshinstall.sh https://mesh.mydomain.lv 'xxxxxxxxxx' `

when i run it on remote site, i got(tried to blank all sensitive info )
![image](https://user-images.githubusercontent.com/60253253/78631509-44f40080-78a5-11ea-8cfd-453ac688d95e.png)

**Comment by Ylianst** [comment]
Ha! Thanks. That screen shot helped, The fix will be in MeshCentral v0.5.1-j when it is published.

**Comment by Ylianst** [comment]
Just published MeshCentral v0.5.1-j, let me know if it works.

**Comment by uldiseihenbergs** [comment]
Just tested with the same settings, and know i can add new Linux agent without a problem. so thanks!
and according to `"lockagentdownload": true` - it would be great, if we could have this option in the WEB UI. just for better security and with option to switch it on/off without MC restart..

**Comment by Ylianst** [comment]
I will close this issue, if you do need some options to be changeable at runtime. Open a new issue, I will take a look at it.
